### ISSUE TITLE
Integrate optional CavJAX cavity into CPU GOSTSHYP

### ISSUE THREAD

**Issue by maxscheurer** [maxscheurer/pyscf-forge#16]
## Original Request

> Let's stick to the CPU implementation for now. Let's create an issue on `maxscheurer/pyscf-forge` for the integration of cavjax.

## Summary

Integrate [CavJAX](https://github.com/maxscheurer/cavjax) as an optional CPU cavity backend for the `pyscf-forge` GOSTSHYP implementation. CavJAX remains independently usable without PySCF; its planned `MolecularCavity` convenience API owns the modified-Bondi radius lookup and `points_per_atom` translation, while `pyscf-forge` owns GOSTSHYP options, normal conventions, electronic integral contractions, and the complete energy chain rule.

The integration must preserve GOSTSHYP's integral-direct evaluation. Electronic integral blocks are contracted immediately, only compact per-grid surface cotangents are retained, and one CPU CavJAX VJP supplies the nuclear geometry response. GPU4PySCF integration is explicitly deferred.

## Dependency

Implement after [CavJAX issue 1](https://github.com/maxscheurer/cavjax/issues/1), which adds the PySCF-backed, array-based `MolecularCavity` convenience API.

Initial CavJAX source: [`maxscheurer/cavjax`](https://github.com/maxscheurer/cavjax), commit [`7020f0e`](https://github.com/maxscheurer/cavjax/commit/7020f0e).

## Current Behavior

- GOSTSHYP supports `vdw`, `vdw/occ`, and optional MOIST `drop` cavities.
- The MOIST integration on `feature/issue-1-add-moist-drop` includes non-rigid coordinate and normal-response mechanics.
- The workshop CavJAX adapter validates energies and gradients using dense JAX geometry Jacobians, but that path is intentionally non-production and does not scale.
- `_grad_direct()` evaluates electronic integrals incrementally, but currently contracts explicit area, coordinate, and normal derivatives rather than passing compact surface sensitivities to a cavity pullback.

## Proposed User Interface

Add a cavity selection analogous to MOIST:

```python
options = {
    "cavity": "cavjax",
    "cavjax_kwargs": {
        "points_per_atom": 100,
        "target_shape_directions": 162,
        "support_temperature": 0.2834589186938655,  # Bohr
        "wall_temperature": 0.15117808997006163,   # Bohr
        "radius_scale": 1.0,
        "radius_offset": 0.0,                       # Bohr
    },
}
```

The bridge passes `mol.atom_charges()` and these options to CavJAX's `MolecularCavity`. PySCF coordinates are already in Bohr. Exact option names may be refined during implementation, but all length-valued options must remain explicitly documented in Bohr.

## Proposed Surface Construction

For `cavity="cavjax"`:

1. Lazily import CavJAX and raise a clear error if unavailable.
2. Construct one CPU `MolecularCavity` using atomic numbers and `cavjax_kwargs`.
3. Call it with `mol.atom_coords()` in Bohr.
4. Retain the CPU CavJAX object and geometry needed for a later VJP.
5. Consume points in Bohr and areas in Bohr².
6. Negate CavJAX's outward normals once to obtain GOSTSHYP inward normals.
7. Preserve the selected point count and topology across coordinate resets for the same atomic composition/options.
8. Transfer only ordinary surface arrays into the existing GOSTSHYP energy machinery.

## Proposed Direct Gradient

Do not materialize full-cavity `nao² × Q` integral tensors or dense cavity Jacobians.

During `_grad_direct()`, contract electronic integral chunks immediately into:

- explicit AO/basis-center nuclear terms, accumulated directly as `(N, 3)`;
- `dE/dpoints`, shape `(Q, 3)`;
- `dE/dareas`, shape `(Q,)`;
- `dE/dnormals`, shape `(Q, 3)`.

Reuse the normal-response algebra from `feature/issue-1-add-moist-drop` to obtain the uncontracted sensitivity with respect to GOSTSHYP's inward normals. Do not apply MOIST's atom-radial tangent projection for CavJAX; negate the inward-normal cotangent and pass it to CavJAX's outward-normal VJP.

After all integral blocks have been contracted and released, invoke exactly one CPU CavJAX response:

```python
geometry_gradient = molecular_cavity.response(
    mol.atom_coords(),
    SurfaceCotangent(
        points=dE_dpoints,
        areas=dE_dareas,
        normals=-dE_dnormals_inward,
    ),
)
```

Add that response to the explicit nuclear contribution. The full retained surface cotangent contains only `7Q` float64 values (approximately 8.6 MiB at 161,292 points).

The first implementation may support only `direct=True` for CavJAX. If so, `direct=False` must raise a clear unsupported-mode error rather than falling back to dense Jacobians or incomplete derivatives.

## Acceptance Criteria

- CavJAX is optional; importing and using unrelated `pyscf-forge` functionality does not require it.
- Selecting `cavity="cavjax"` without CavJAX installed raises a clear installation error and never silently falls back.
- The adapter uses CavJAX's `MolecularCavity` convenience API rather than duplicating modified-Bondi radius lookup or `points_per_atom` translation.
- Coordinates and all length-valued options are passed in Bohr; areas are consumed as Bohr².
- CavJAX outward normals are converted to GOSTSHYP inward normals exactly once.
- Cavity reset preserves fixed point count/topology for unchanged atomic composition and options and invalidates all dependent GOSTSHYP caches.
- The direct gradient does not materialize dense cavity geometry Jacobians.
- The direct gradient does not materialize full-cavity `nao² × Q` integral tensors; integral blocks are contracted and released incrementally.
- The only full-grid derivative intermediate introduced for CavJAX is a compact point/area/normal cotangent of total size `7Q`.
- Exactly one CPU CavJAX VJP is used per GOSTSHYP gradient evaluation.
- Point, area/width, and normal contributions are all included with tested units and signs.
- Unsupported non-direct CavJAX gradient modes fail clearly rather than using an incorrect fallback.
- A small closed-shell molecule and a small intermolecular cluster pass converged-SCF finite-difference gradient checks.
- Translation invariance and gradient force-sum checks pass within documented tolerances.
- Existing `vdw`, `vdw/occ`, and `drop` behavior remains unchanged.
- Cavity construction, compact-cotangent preparation, and CavJAX VJP timings are recorded separately from electronic integral timings.
- Documentation covers optional installation, all CavJAX-specific controls, Bohr units, global versus per-atom point budgets, direct-mode support, and known shape limitations.
- The normal test suite passes without CavJAX installed, and the installed-backend integration tests run in standard CI now that CavJAX is public.

## Implementation Plan

1. Complete the `MolecularCavity` API tracked by CavJAX issue 1.
2. Add a lazy optional-dependency bridge under `pyscf/solvent`, analogous to `pyscf/solvent/moist.py`.
3. Add CavJAX-specific option parsing, validation, logging, build/reset state, and surface extraction.
4. Refactor the CPU direct gradient algebra to expose compact point, area, and inward-normal cotangents while retaining immediate integral contraction.
5. Reuse the MOIST branch's normal sensitivity before its atom-radial derivative contraction.
6. Convert the normal cotangent to outward convention and call one CPU CavJAX VJP.
7. Validate against the workshop dense-Jacobian adapter, then retain no production dependency on that path.
8. Add optional-dependency, surface, reset, chain-rule, finite-difference, and timing tests.
9. Document configuration and limitations.

## Known Model Limitations

Document rather than conceal CavJAX's current finite-direction orientation dependence, smoothing inflation, strict global convexity, and non-equal-area adapted sampling. Software integration and gradient correctness do not by themselves establish physical convergence.

## CI and Packaging

CavJAX is public, so installed-backend tests can run in standard CI without repository credentials. Until a released package version is available, development CI may install a pinned Git commit; release metadata should ultimately use a normal optional dependency rather than an unpinned branch.

## Out of Scope

- GPU4PySCF integration
- Running CavJAX geometry or VJPs on GPU
- Dense CavJAX geometry Jacobians
- Reimplementing PySCF radii or point-budget convenience logic in `pyscf-forge`
- Reintroducing rejected workshop `support` or `radial` modes
- Changing existing GOSTSHYP amplitude or denominator policies
- Declaring the current cavity universally physically validated

**Comment by maxscheurer** [comment]
## Issue Assessment

### Summary

Integrate CavJAX as a lazy, optional, CPU-only cavity backend for GOSTSHYP while preserving the existing integral-direct electronic evaluation. The central work is not surface extraction: it is refactoring the direct gradient so chunked libcint results produce one explicit nuclear contribution plus compact point, area, and inward-normal cotangents, followed by one CavJAX pullback.

The prerequisite is complete. CavJAX issue 1 was closed by merged pull request 2 at commit `b2d0aa8`, which released the repository state as version `0.2.0` and added the required `MolecularCavity` API.

### Current state

- `pyscf/solvent/gostshyp.py` supports `vdw`, `vdw/occ`, and `drop`. `build()` dispatches the backend, derives Gaussian widths and normalizations, and invalidates the cached `gtilde` and `force_operators` tensors.
- `pyscf/solvent/moist.py` provides the nearest optional-backend pattern: deferred use from GOSTSHYP, a guarded dependency import, a clear installation error, surface extraction, and geometry derivatives.
- GOSTSHYP normals are inward. The standard paths derive them from an owning atom; CavJAX returns outward normals and its global convex surface has no physical point owner, so the CavJAX build path must supply negated normals directly rather than pass through `compute_surface_normals`.
- `_grad_direct()` already chunks the large libcint tensors and contracts most of them promptly. It currently consumes dense `dareas`, `dcoords`, and `dnormals` tensors, however, rather than accumulating surface cotangents.
- `_normal_derivative_term()` contains the needed uncontracted normal sensitivity: it forms `P_gc` with shape `(Q, 3)` and multiplies it by `dE/dF`. It currently builds the full `nao² × Q × 3` p-type integral outside the chunk loop, so it cannot be reused unchanged under this issue's memory criteria.
- The local workshop adapter under `scratch/cavity_workshop/` validates the surface sign/unit bridge and converged-SCF gradients with dense JAX Jacobians. It is untracked, explicitly marked as a validation-only path, and documents the pending contracted-VJP stage.
- Existing tests cover direct-versus-cached behavior, finite-difference gradients, spherical and Cartesian bases, multi-atom systems, translation invariance, reset, and optional MOIST behavior. No production CavJAX bridge or tests exist yet.
- Packaging has an optional-extra precedent in `setup.py`, but current CI installs neither MOIST nor CavJAX. The CI matrix includes Python 3.8, 3.10, and 3.12.
- CavJAX's shipped API is `MolecularCavity.build(positions)`, not a callable object. `response(positions, SurfaceCotangent(...))` is present. It is Bohr-native, uses outward normals, requires JAX float64, fixes topology at construction, and lazily loads PySCF radii when `MolecularCavity` is constructed.

### Gaps

- There is no CavJAX bridge, option parsing, cavity dispatch, dependency error, surface validation, reset state, timing split, package extra, documentation, or installed-backend CI path.
- `build()` always recomputes atom-radial normals after backend construction. It needs a backend-aware path that accepts CavJAX normals exactly once and does not invent point ownership.
- The direct gradient must be reorganized to accumulate `dE/dpoints`, `dE/dareas`, and `dE/dnormals` while preserving separate AO/basis-center nuclear terms. Area sensitivity is distributed across the explicit pressure-area term and the d- and f-type width terms; point sensitivity is distributed across s- and p-type Gaussian-center terms; normal sensitivity comes from the force operator. All must be retained with the current signs and normalization factors.
- The normal sensitivity must move into the existing chunk loop. Otherwise the current full-grid p-type integral violates the stated integral-direct bound even if the final CavJAX response is a VJP.
- The CavJAX path needs an explicit unsupported-mode contract for `direct=False`; the current cached route necessarily materializes full-cavity `nao² × Q` operators.
- Reset currently rebuilds the surface and clears major integral/model state, but the CavJAX object, fixed atomic identities/options, geometry, cotangents, response state, and new timing fields need explicit lifecycle rules and invalidation.
- Tests must separately establish optional-import behavior, surface units/signs, topology preservation, compact-cotangent equivalence to the dense workshop reference, exactly one response call, finite-difference total gradients, invariance, timing fields, and no regressions in existing cavities.

### Ambiguities to settle during planning

1. **Dependency pin and release source.** CavJAX has merged version `0.2.0` at `b2d0aa8` but currently has no GitHub release. Decide whether the initial extra/CI installs that exact commit and what minimum released version will replace it.
2. **Python CI split.** CavJAX requires Python 3.10 or newer while this repository tests Python 3.8. Installed-backend tests therefore need a version-gated CI job; the Python 3.8 job should continue proving that ordinary installation and tests do not require CavJAX.
3. **CPU enforcement.** CavJAX uses JAX's selected default device and does not itself force CPU placement. Define how this bridge guarantees CPU construction and response on hosts where a GPU-enabled JAX installation is present, without mutating process-global JAX configuration unexpectedly.
4. **Meaning of `direct=False`.** Clarify whether CavJAX construction rejects it immediately, whether energy evaluation may proceed but `grad()` rejects it, or whether only an attempted non-direct gradient raises. Early rejection is simpler, but the issue currently specifies only that unsupported gradient modes fail clearly.
5. **Reset with changed composition.** `MolecularCavity` fixes atomic numbers and radii at construction. Specify whether a reset with changed identities/order creates a new object or raises and requires a new GOSTSHYP instance. Unchanged composition should reuse the object and fixed topology.
6. **Timing semantics.** Define the public field names and whether construction timing includes first JAX compilation. Also clarify whether values are per-call or accumulated; current kernel time accumulates while gradient time is overwritten.
7. **The `7Q` memory statement.** Clarify that this bounds newly retained full-grid derivative arrays, not ordinary surface/model vectors, chunk-local electronic tensors, or internal JAX pullback workspaces.
8. **Point ownership compatibility.** The global CavJAX surface has no owner array. Decide whether internal code should allow `atom_idx=None` for this backend or retain an explicitly inert compatibility array; no gradient path may silently use rigid owner scattering.

### Scope and complexity

**Large, concentrated numerical refactor.** The optional bridge and surface setup are moderate, but extracting three complete cotangents from the existing direct-gradient algebra while preserving legacy paths is high-risk work. Expected changes include one bridge module, substantial edits to `gostshyp.py`, focused unit and integration tests, packaging/CI changes, and documentation updates.

A safe implementation should use the dense workshop Jacobian path only as a test oracle, compare each point/area/normal contribution independently, and keep the production implementation exclusively on the one-VJP path.

### Risks and edge cases

- Sign mistakes can arise at three boundaries: libcint Gaussian-center derivative conventions, GOSTSHYP inward normals, and CavJAX outward-normal cotangents. Negating both the surface normal and its cotangent in the wrong place can produce plausible but incorrect total gradients.
- Area derivatives affect Gaussian widths and normalizations in multiple terms. Collecting only the explicit pressure-area term would produce an incomplete chain rule.
- Refactoring shared direct-gradient code can regress `vdw`, `vdw/occ`, or `drop`; preserving those paths and their direct-versus-cached reference tests is essential.
- Negative-amplitude masking is nondifferentiable when a perturbation crosses the mask threshold. Finite-difference fixtures should report denominator/amplitude diagnostics and avoid interpreting threshold crossings as cavity-response failures.
- JAX x64 must be enabled before cavity construction. The bridge should surface CavJAX's error clearly and avoid silently running float32.
- First-call JIT compilation can dominate timing and can complicate assertions that exactly one VJP occurs. Tests should count the public `response` invocation independently from compilation internals.
- A fixed global point budget does not imply atom-centered ownership or equal-area sampling. Documentation must preserve that distinction and the model limitations already listed in the issue.
- Although the external cotangent is only `7Q`, the JAX pullback may require substantial internal workspace for large point and shape-direction counts. C60-scale memory and timing should be measured rather than inferred from cotangent size alone.
- Optional import tests must prove both sides: unrelated GOSTSHYP works without CavJAX, while selecting `cavjax` fails clearly and never falls back.

### Assessment

The issue is ready for detailed planning: its prerequisite API now exists and the workshop reference provides a strong correctness oracle. The main planning deliverable should be a term-by-term direct-gradient decomposition that identifies the explicit nuclear and three surface-cotangent accumulators before implementation begins. The dependency pin, CPU-placement rule, non-direct failure point, reset behavior for changed composition, and timing semantics should be fixed in that plan.

---
*Automated assessment by mach6*
